### PR TITLE
[v2] Add upgrade action retry

### PR DESCRIPTION
**Pull request by michel-laterman**
## What does this PR do?

Add the ability for the agent to schedule and retry upgrade actions.

The fleetapi actions now define a ScheduledAction, and RetryableAction interface to eliminate the need for stub methods on all different action types. Action queue has been changed to function on scheduled actions. Serialization tests now ensure that that the retry attribute needed by retryable actions works.

Decouple dispatcher from gateway, dispatcher has an errors channel that will return an error for the list of actions that's sent. Gateway has an Actions method that can be used to get the list of actions from the gateway. The managed_mode config manager will link these two components

If a handler returns an error and the action is a RetryableAction, the dispatcher will attempt to schedule a retry. The dispatcher will also ack the action to fleet-server and indicate if it will be retried or has failed (or has been received normally). 
For the acker, if a RetryableAction has an error and an attempt count that is greater than 0 it will be acked as retried. If it has an error and an attempt count less than 1 it will be acked as failed.

## Why is it important?

In order better support upgrades at scale the agent should be able to retry by itself in case of a temporary issue, such as network failure.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

TODO

## Related issues

- Relates #778

### CODE REVIEW

**Comment by jlind23** [comment]
@michel-laterman would it be ready to be merge in 8.5 or should we postpone it?

**Comment by michel-laterman** [comment]
@jlind23, holding for 8.6 - it also needs a fleet PR

**Comment by michel-laterman** [comment]
Added a note to the the upgrade docs: https://github.com/elastic/observability-docs/pull/2230

File: internal/pkg/agent/application/dispatcher/dispatcher.go
Comment (blakerouse) [inline internal/pkg/agent/application/dispatcher/dispatcher.go:154]: Why use a multierror here? Why not just sent the error over the errCh, each time one occurs?

File: internal/pkg/agent/application/dispatcher/dispatcher.go
Comment (blakerouse) [inline internal/pkg/agent/application/dispatcher/dispatcher.go:140]: Still don't like the usage of custom error type. What about any error results in it being triggered for retry if its retryable?

File: internal/pkg/agent/application/gateway/fleet/fleet_gateway.go
Comment (blakerouse) [inline internal/pkg/agent/application/gateway/fleet/fleet_gateway.go:165]: This is great. I really like the separation between the 2.

File: internal/pkg/agent/application/managed_mode.go
Comment (blakerouse) [inline internal/pkg/agent/application/managed_mode.go:149]: I believe this whole block of code is blocking. `actionDispatcher.Dispatch` is a blocking call, its not running in a goroutine so that means the whole content of the function needs to complete before running. But if an error happens inside of the code block it will try to write to the `errCh` which will block because nothing is reading from it.

File: internal/pkg/agent/application/managed_mode.go
Comment (blakerouse) [inline internal/pkg/agent/application/managed_mode.go:209]: You should just start this before restoring the policy from disk. Then this gofunction will handle reading from the `errCh` and sending it to the `errCh` for the managed mode.

File: internal/pkg/agent/application/managed_mode.go
Comment (blakerouse) [inline internal/pkg/agent/application/managed_mode.go:224]: I like this design much better. Separating the 2 and then connect them together in the manager.

File: internal/pkg/agent/application/upgrade/upgrade.go
Comment (blakerouse) [inline internal/pkg/agent/application/upgrade/upgrade.go:164]: This doesn't seem like something we need to place everywhere. Because as soon as a developer forgets to add that error they no longer get retried. I would just say if an action errors and its retryable we retry it.

**Review by blakerouse — CHANGES_REQUESTED** [review]
Overall I really like how this operates. There are a few things that need to be looked and and fixed.

File: internal/pkg/agent/application/upgrade/upgrade.go
Comment (michel-laterman) [reply internal/pkg/agent/application/upgrade/upgrade.go:164]: Do we want to consider all errors retryable? We know that `ErrSameVersion` will never succeed.

File: internal/pkg/agent/application/managed_mode.go
Comment (michel-laterman) [reply internal/pkg/agent/application/managed_mode.go:149]: good catch, i'll make the changes

File: internal/pkg/agent/application/dispatcher/dispatcher.go
Comment (michel-laterman) [reply internal/pkg/agent/application/dispatcher/dispatcher.go:154]: 👍 this was done before i made the error channel for the dispatcher. have removed

File: internal/pkg/agent/application/managed_mode.go
Comment (blakerouse) [inline internal/pkg/agent/application/managed_mode.go:159]: Why is this still here? Isn't the new goroutine started above going to read and handle these errors?

File: internal/pkg/agent/application/upgrade/upgrade.go
Comment (blakerouse) [inline internal/pkg/agent/application/upgrade/upgrade.go:145]: Is this really an error and did it really fail? I would prefer to see:

```suggestion
		u.log.Warn("Upgrade action skipped: upgrade did not occur because its the same version")
```

**Review by blakerouse — CHANGES_REQUESTED** [review]
Thanks for the changes and all the testing. I think this looks really good and is basically ready to land. I just have one question that I am going to block the PR on, because it seems incorrect and should be removed (unless I am misunderstanding).

File: internal/pkg/agent/application/managed_mode.go
Comment (michel-laterman) [reply internal/pkg/agent/application/managed_mode.go:159]: changed

**Review by kpollich — APPROVED** [review]
Thanks for all the comments and helpful test cases 🚀 

I read through Blake's comments here and things look good from my end. I'll defer to him as more of a subject matter expert here but overall awesome work here @michel-laterman. Thanks for reaching out about this.

**Review by blakerouse — APPROVED** [review]
Thanks for working on this and the fixes! Looks great now!

**Pull request merged**
